chain/starknet/src/data_source.rs

```diff
@@ -0,0 +1,320 @@
+    trigger::{StarknetEventTrigger, StarknetTrigger},
+};
+
+const BLOCK_HANDLER_KIND: &str = "block";
+const EVENT_HANDLER_KIND: &str = "event";
+
+#[derive(Clone)]
+pub struct DataSource {
+    pub kind: String,
+    pub network: String,
+    pub name: String,
+    pub source: Source,
+    pub mapping: Mapping,
+}
+
+#[derive(Clone)]
+pub struct Mapping {
+    pub block_handlers: Vec<MappingBlockHandler>,
+    pub event_handlers: Vec<MappingEventHandler>,
+    pub runtime: Arc<Vec<u8>>,
+}
+
+#[derive(Deserialize)]
+pub struct UnresolvedDataSource {
+    pub kind: String,
+    pub network: String,
+    pub name: String,
+    pub source: Source,
+    pub mapping: UnresolvedMapping,
+}
+
+#[derive(Clone, PartialEq, Eq, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Source {
+    pub start_block: BlockNumber,
+    #[serde(default, deserialize_with = "deserialize_address")]
+    pub address: Option<FieldElement>,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct UnresolvedMapping {
+    #[serde(default)]
+    pub block_handlers: Vec<MappingBlockHandler>,
+    #[serde(default)]
+    pub event_handlers: Vec<UnresolvedMappingEventHandler>,
+    pub file: Link,
+}
+
+#[derive(Clone, PartialEq, Eq, Deserialize)]
+pub struct MappingBlockHandler {
+    pub handler: String,
+}
+
+#[derive(Clone, PartialEq, Eq, Deserialize)]
+pub struct MappingEventHandler {
+    pub handler: String,
+    pub event_selector: FieldElement,
+}
+
+#[derive(Clone, Deserialize)]
+pub struct UnresolvedMappingEventHandler {
+    pub handler: String,
+    pub event: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct DataSourceTemplate;
+
+#[derive(Clone, Default, Deserialize)]
+pub struct UnresolvedDataSourceTemplate;
+
+impl blockchain::DataSource<Chain> for DataSource {
+    fn from_template_info(_template_info: DataSourceTemplateInfo<Chain>) -> Result<Self, Error> {
+        Err(anyhow!("StarkNet subgraphs do not support templates"))
+    }
+
+    fn address(&self) -> Option<&[u8]> {
+        None
+    }
+
+    fn start_block(&self) -> BlockNumber {
+        self.source.start_block
+    }
+
+    fn handler_kinds(&self) -> HashSet<&str> {
+        let mut kinds = HashSet::new();
+
+        let Mapping {
+            block_handlers,
+            event_handlers,
+            ..
+        } = &self.mapping;
+
+        if !block_handlers.is_empty() {
+            kinds.insert(BLOCK_HANDLER_KIND);
+        }
+        if !event_handlers.is_empty() {
+            kinds.insert(EVENT_HANDLER_KIND);
+        }
+
+        kinds
+    }
+
+    fn match_and_decode(
+        &self,
+        trigger: &StarknetTrigger,
+        block: &Arc<codec::Block>,
+        _logger: &Logger,
+    ) -> Result<Option<TriggerWithHandler<Chain>>, Error> {
+        if self.start_block() > block.number() {
+            return Ok(None);
+        }
+
+        let handler = match trigger {
+            StarknetTrigger::Block(_) => match self.mapping.block_handlers.first() {
```

Comment:
If there can only be 1 block handler maybe we could change this from a vec to a single handler? Including in the validations?

Reply:
I think this was originally copied from the `near` implementation. The `near` code has been refactored a bit but you can still see this line taking only the first block handler:

https://github.com/graphprotocol/graph-node/blob/53dbe9356ff4515cee08f9a0b92c70161dee11fa/chain/near/src/data_source.rs#L304-L306

and in turn called in `match_and_decode()`:

https://github.com/graphprotocol/graph-node/blob/53dbe9356ff4515cee08f9a0b92c70161dee11fa/chain/near/src/data_source.rs#L139-L142

I'm not sure why it's implemented this way, but I figured I should better follow suit.

Reply:
yeah I saw near does the same, I also don't understand why that decision was made. The main issue is that if this is not validated it's easy for people to think they can define more than one handler and then it will just not run. The UnresolvedMapping has a `block_handlers: Vec...` property which will allow someone to define multiple handlers, if that turns into a single String handler then I think this is fine to keep internally if you prefer that way, although it would be better to just use Option internally in that case.

Reply:
Hmm for the unresolved part I can totally see that we can't really fix it for chains like `near` without bumping manifest version. For `starknet` though we can totally fix it, both the unresolved and resolved types. Do you prefer if I fix it now and use `Option` for both unresolved and resolved?

Reply:
We can merge the PR for now so there's less of a chance for conflicts and then you can fix and add some tests on a separate PR, you can check near/chain.rs's tests, just making sure we have some examples in order to avoid breaking things

Reply:
Sounds good! Will definitely follow up with another PR on this. As long as we do it before the next release we aren't really breaking anyone after all.